File: leetcode/Combination_Sum_2.cpp
class Solution {
public:
    vector<vector<int> > combinationSum2(vector<int> &num, int target) {
        vector<vector<int>> result;
        sort(num.begin(), num.end());
        
        combine_helper(result, num, 0, target, vector<int>());
        
        return result;
    }
private:
    void combine_helper(vector<vector<int>> &result, vector<int> &num, int s, int target, vector<int> sofar){
        if(target == 0){
            result.push_back(sofar);
            return;
        }
        
        if(s >= num.size()){
            return;
        }
        
        int i = s;
        while(i < (int)num.size()-1 && num[i] == num[i+1]) i++;
        
        // not choose any element
        combine_helper(result, num, i+1, target, sofar);
        
        for(int k = s; k <= i; k++){
            sofar.push_back(num[s]);
            target -= num[s];
            if(target < 0) return;
            
            combine_helper(result, num, i+1, target, sofar);
        }
    }
};

File: leetcode/java/CombinationSum_2.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        if(candidates == null || candidates.length == 0) return Collections.emptyList();
        
        Arrays.sort(candidates);
        List<List<Integer>> result = new ArrayList<>();
        csh(result, new ArrayList<>(), 0, candidates, 0, target);
        return result;
    }
    
    
     private void csh(List<List<Integer>> result, List<Integer> sofar, int sum, int[] cands, int s, int t) {
        if(sum == t) {
            result.add(new ArrayList<>(sofar));
            return;
        }
        if(s == cands.length) return;

        int i = s;
        while(i < cands.length-1) {
            if(cands[i] != cands[i+1]) break;
            i++;
        }
        i++;

        int x = cands[s];
        int k = s;
        int count = 0;
        while(k <= i) {
            if(sum + (k-s)*x > t) break;
            if(k > s) {
                sofar.add(x);
                count++;
            }
            csh(result, sofar, sum+(k-s)*x, cands, i, t);
            k++;
        }

        while(count > 0) {
            sofar.remove(sofar.size()-1);
            count--;
        }
    }
}
